internal/pkg/test/privilege_linux.go

```diff
@@ -77,8 +79,7 @@ func ResetPrivilege(t *testing.T) {
 
 	runtime.UnlockOSThread()
 
-	// set SINGULARITY_CACHEDIR
-	os.Setenv("SINGULARITY_CACHEDIR", CacheDirPriv)
+	os.Setenv("SINGULARITY_TESTBASEDIR", BaseDirPriv)
```

Comment:
This is making the assumption that `origUID`/`origGID` is a privileged user, but this isn't _technically_ correct. Certain package tests can be run unprivileged, and it is useful to be able to do that when, for example, running tests directly from VS Code rather than the command line.

Reply:
Okay but I am sure that I am missing something because this is the assumption by the original code, not my modification, right? `ResetPrivilege()` is meant to be called after `DropPrivilege()` initially in the context of a privilege process. Furthermore, if not executed in the context of a privileged context in the first place, the behavior following the modification will still be the same.

I am a little confused here: the intent of this PR is not to revisit how privileged/unprivileged tests are executed but simply to avoid concurrent and unwanted sharing of caches which are creating problems.

Reply:
> ResetPrivilege() is meant to be called after DropPrivilege()

Yes, this is true.

> ...initially in the context of a privilege process.

This assumption is not necessarily true. `ResetPrivilege()` sets effective privilege to the _original_ user, which could be privileged or unprivileged. Many packages can be tested as an unprivileged user, and this is common when working within a package using an editor such as VS Code.

> I am a little confused here: the intent of this PR is not to revisit how privileged/unprivileged tests are executed but simply to avoid concurrent and unwanted sharing of caches which are creating problems.

Yes, but `SetCacheDir` and `CleanCacheDir` rely on side-effects of functions such as `ResetPrivilege`, and thus they must be considered. Consider the following sequence of events:

- `/tmp/WithPrivilege` directory exists and is writable by `root` only.
- `go test github.com/sylabs/singularity/blah/blah` is run as unprivileged user, on command line or via editor.
- one of the tests in the package calls `DropPrivilege()` and then `ResetPrivilege()`. These are effectively no-ops, since the test was run as an unprivileged user. `ResetPrivilege()` `SINGULARITY_TESTBASEDIR` to `/tmp/WithPrivilege`.
- a test calls `SetCacheDir()`, which attempts to call `ioutil.TempDir()` on `/tmp/WithPrivilege`, which fails with an access denied error.

Reply:
Sorry it seems that I am not precise enough here, my intent was not to talk about a general case. And I start to believe that we ultimately are saying the same thing and we just identified a missing piece.

I have the following code path in mind, which happens all the time when manually running `make testall`:
- A unit test in privileged creates a new temporary image cache for testing, usually `EnsurePrivilege()` is called; `SINGULARITY_CACHEDIR` is set to point at this new image cache.
- The unit test switches to unprivileged mode by calling `DropPrivilege()`.
- After switching to unprivileged mode, a new image cache is usually *not* set with the current implementation of the unit test **and** the environment variable still points at the one that was created In privileged mode because `SINGULARITY_CACHEDIR` is set to point to it.
- The unit test is calling functions that ultimately rely on the image cache (it is usually not explicit, more like a side effect) and if the cache is not set to be shared between privileged and unprivileged mode, the test fails.

I believe the key point here is `SINGULARITY_TESTBASEDIR`. That concept was not there before and I actually removed it because I thought you said it was not what you had in mind, while I know believe it was just not implemented correctly to reflect what you had in mind.

So here is how I would phrase the semantics we are looking for here and let's see if we are in agreement:
`SINGULARITY_DEFAULT_TESTBASEDIR` (a.k.a `SINGULARITY_TESTBASEDIR` but i renamed it to reflect better what it means) sets the default location that we use as base directory to set temporary image caches for unit tests. Unit test developers can override that location by specifying the base directory when calling `SetCacheDir()`. When switching between privileged and unprivileged mode within a unit test (by calling functions from `internal/pkg/test`), `SINGULARITY_DEFAULT_TESTBASEDIR` will automatically be set to the appropriate default location, i.e., `/tmp/WithPrivilege` or `/tmp/WithoutPrivilege`).
It is the developer's responsibility to correctly initialize a new temporary image cache, as appropriate, when switching between privileged and unprivileged mode, by properly calling `SetCacheDir()` and properly setting `SINGULARITY_CACHEDIR` for the temporary image cache to be used by the unit test. It is also the developer's responsibility to correctly delete all temporary image caches that have been set in the context of unit tests.

Reply:
I still wonder if we need `SINGULARITY_DEFAULT_TESTBASEDIR` since `SetCacheDir()` will either rely on the path that is passed in or simply create a new temporary directory.

I guess my question would be: do we want to have all the temporary image caches in `/tmp/WithPrivilege` and `/tmp/WithoutPrivilege` by default? or just start to put them wherever `ioutil.TempDir("", ...)` would put them?